### PR TITLE
feat: overwrite teav value if already exist in original

### PR DESCRIPTION
**Pull request by ameenhere**
Overwrites existing teav value if already exists in original tei.
If not, a new teav is created and attached to original tei.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java [hunk 1]
```diff
@@ -165,24 +167,45 @@ public boolean exists( PotentialDuplicate potentialDuplicate )
     public void moveTrackedEntityAttributeValues( TrackedEntityInstance original, TrackedEntityInstance duplicate,
         List<String> trackedEntityAttributes )
     {
+        // Collect existing teav from original for the tea list
+        Map<String, TrackedEntityAttributeValue> originalAttributeValueMap = new HashMap<>();
+        original.getTrackedEntityAttributeValues().forEach( oav -> {
+            if ( trackedEntityAttributes.contains( oav.getAttribute().getUid() ) )
+            {
```
Comment (lucaCambi77) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java:174]: Not sure about the standards here, but can we go like 

```
        Map<String, TrackedEntityAttributeValue> originalAttributeValueMap = new HashMap<String, TrackedEntityAttributeValue>() {
            {
                original.getTrackedEntityAttributeValues().forEach(oav -> {
                    if (trackedEntityAttributes.contains(oav.getAttribute().getUid())) {
                        put(oav.getAttribute().getUid(), oav);
                    }
                });
            }
        };
```
In any case i'd leave HashMap<String, TrackedEntityAttributeValue> for java 8 compatibility

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java [hunk 2]
```diff
@@ -165,24 +167,45 @@ public boolean exists( PotentialDuplicate potentialDuplicate )
...
             .filter( av -> trackedEntityAttributes.contains( av.getAttribute().getUid() ) )
             .forEach( av -> {
-                TrackedEntityAttributeValue nav = new TrackedEntityAttributeValue();
-
-                nav.setAttribute( av.getAttribute() );
-                nav.setEntityInstance( original );
-                nav.setValue( av.getValue() );
 
+                TrackedEntityAttributeValue updatedTeav = null;
+                if ( originalAttributeValueMap.containsKey( av.getAttribute().getUid() ) )
```
Comment (lucaCambi77) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java:185]: I think null is redunant here

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java [hunk 3]
```diff
@@ -165,24 +167,45 @@ public boolean exists( PotentialDuplicate potentialDuplicate )
...
+                }
+                else
+                {
+                    // teav does not exist in original, so create new and attach
+                    // it to original
+                    updatedTeav = new TrackedEntityAttributeValue();
+                    updatedTeav.setAttribute( av.getAttribute() );
+                    updatedTeav.setEntityInstance( original );
+                    updatedTeav.setValue( av.getValue() );
+                }
```
Comment (lucaCambi77) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java:199]: Can we move this in a private helper method?

**Review by lucaCambi77 — APPROVED** [review]
Looks good to me, minor comments

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java (hunk 1 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java:174]: @lucaCambi77 Wdym Java 8 compatibility? This type of assignment type inference is supported even in Java 7. So surely in Java 8 as well.
Map<String, TrackedEntityAttributeValue> originalAttributeValueMap = new HashMap<>();

Also regarding the other anonymous class way of initializing hashmaps, I dont think we have followed that way in most cases, so sticking to the plain old technique.

**Review by stian-sandvold — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java (hunk 1 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/hibernate/HibernatePotentialDuplicateStore.java:174]: @lucaCambi77 Wdym Java 8 compatibility? This type of assignment type inference is supported even in Java 7. So surely in Java 8 as well.
Map<String, TrackedEntityAttributeValue> originalAttributeValueMap = new HashMap<>();

Also regarding the other anonymous class way of initializing hashmaps, I dont think we have followed that way in most cases, so sticking to the plain old technique. That technique also creates an extra anonymous class at every usage, so the traditional approach is alright imo.

**Pull request merged**